src/main/java/cn/nukkit/network/protocol/MoveEntityDeltaPacket.java

```diff
@@ -28,49 +46,104 @@ public byte pid() {
-        putCoordinate(FLAG_HAS_Y, this.y);
-        putCoordinate(FLAG_HAS_Z, this.z);
-        putRotation(FLAG_HAS_YAW, this.yawDelta);
-        putRotation(FLAG_HAS_HEAD_YAW, this.headYawDelta);
-        putRotation(FLAG_HAS_PITCH, this.pitchDelta);
+        this.reset();
+        this.putEntityRuntimeId(this.runtimeEntityId);
+        if (flags != 0) {
+            this.putLShort(flags);
+            if ((this.flags & FLAG_HAS_X) != 0) {
+                this.putLFloat(this.x);
+            }
+            if ((this.flags & FLAG_HAS_Y) != 0) {
+                this.putLFloat(this.y);
+            }
+            if ((this.flags & FLAG_HAS_Z) != 0) {
+                this.putLFloat(this.z);
+            }
+            if ((this.flags & FLAG_HAS_PITCH) != 0) {
+                this.putByte((byte) (this.pitch / (360F / 256F)));
+            }
+            if ((this.flags & FLAG_HAS_YAW) != 0) {
+                this.putByte((byte) (this.yaw / (360F / 256F)));
+            }
+            if ((this.flags & FLAG_HAS_HEAD_YAW) != 0) {
+                this.putByte((byte) (this.headYaw / (360F / 256F)));
+            }
+        } else {
+            List<Consumer<MoveEntityDeltaPacket>> consumers = new ArrayList<>(6);
+            if (this.x != 0) {
+                this.flags |= FLAG_HAS_X;
+                consumers.add(packet -> packet.putLFloat(this.x));
+            }
+            if (this.y != 0) {
+                this.flags |= FLAG_HAS_Y;
+                consumers.add(packet -> packet.putLFloat(this.y));
+            }
+            if (this.z != 0) {
+                this.flags |= FLAG_HAS_Z;
+                consumers.add(packet -> packet.putLFloat(this.z));
+            }
+            if (this.pitch != 0) {
+                this.flags |= FLAG_HAS_PITCH;
+                consumers.add(packet -> packet.putByte((byte) (this.pitch / (360F / 256F))));
+            }
+            if (this.yaw != 0) {
+                this.flags |= FLAG_HAS_YAW;
+                consumers.add(packet -> packet.putByte((byte) (this.yaw / (360F / 256F))));
+            }
+            if (this.headYaw != 0) {
+                this.flags |= FLAG_HAS_HEAD_YAW;
+                consumers.add(packet -> packet.putByte((byte) (this.headYaw / (360F / 256F))));
+            }
+            this.putLShort(this.flags);
+            if (consumers.size() > 0) {
+                for (Consumer<MoveEntityDeltaPacket> consumer : consumers) {
+                    consumer.accept(this);
+                }
+            }
+        }
     }
 
-    private float getCoordinate(int flag) {
-        if ((flags & flag) != 0) {
-            return this.getLFloat();
-        }
-        return 0;
+    public float getCoordinate() {
+        return this.getLFloat();
     }
 
-    private double getRotation(int flag) {
-        if ((flags & flag) != 0) {
-            return this.getByte() * (360d / 256d);
-        }
-        return 0d;
+    public void putCoordinate(float value) {
```

Comment:
These should not be public as they are only to be used for internally reading and writing the network packet.

Reply:
yeah